core/src/main/java/org/testcontainers/containers/TestContainer.java

```diff
@@ -0,0 +1,162 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.model.Info;
+import lombok.NonNull;
+import org.testcontainers.containers.output.OutputFrame;
+import org.testcontainers.containers.traits.LinkableContainer;
+import org.testcontainers.containers.wait.WaitStrategy;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.time.Duration;
+import java.util.List;
+import java.util.function.Consumer;
+
+public interface TestContainer<SELF extends TestContainer<SELF>> extends LinkableContainer {
+
+    default SELF self() {
+        return (SELF) this;
```

Comment:
the only way to self typing it in Java is this line :( anyway, it's safe because of convention)